man/scholdoc-citeproc.1.md
---
title: scholdoc-citeproc
section: 1
...

# NAME

scholdoc-citeproc - filter to resolve citations in a Pandoc/Scholdoc document.

# SYNOPSIS

scholdoc-citeproc [options] [file..]

# DESCRIPTION

Scholdoc-citeproc is a fork of `pandoc-citeproc`, meant to be used with
Scholdoc instead of Pandoc.

The `scholdoc-citeproc` executable has two modes, filter mode and convert mode.

## Filter mode

Run without options, it acts as a filter that takes a JSON-encoded
Pandoc/Scholdoc document, formats citations and adds a bibliography, and
returns a JSON-encoded scholdoc document. Citations will be resolved, and a
bibliography will be appended to the end of the document (unless the
`suppress-bibliography` metadata field is set to a true value). If you wish the
bibliography to have a section header, put the section header at the end of
your document. (See the `pandoc_markdown` (5) man page under "Citations" for
details on how to encode citations in pandoc's markdown.)

To process citations with scholdoc, call scholdoc-citeproc as a filter:

    scholdoc --filter scholdoc-citeproc input.md -s -o output.html

The bibliography will be put into a pandoc `Div` container with
class `references`.

scholdoc-citeproc will look for the following metadata fields in
the input:

`bibliography`:  A path, or YAML list of paths, of bibliography
files to use.  These may be in any of the formats supported by
bibutils.

    Format            File extension
    ------------      --------------
    MODS              .mods
    BibLaTeX          .bib
    BibTeX            .bibtex
    RIS               .ris
    EndNote           .enl
    EndNote XML       .xml
    ISI               .wos
    MEDLINE           .medline
    Copac             .copac
    JSON citeproc     .json

`references`:  A YAML list of references.  Each reference is a YAML
object.  The format is essentially CSL JSON format.  Here is an example:

    - id: doe2006
      author:
        family: Doe
        given: [John, F.]
      title: Article
      page: 33-34
      issued:
        year: 2006
      type: article-journal
      volume: 6
      container-title: Journal of Generic Studies

The contents of fields will be interpreted as markdown when
appropriate:  so, for example, emphasis and strong emphasis can
be used in title fileds. Simple tex math will also be
parsed and rendered appropriately.

`csl` or `citation-style`: Path or URL of a CSL style file.
If the file is not found relative to the working directory,
scholdoc-citeproc will look in the `$HOME/.csl` directory (or
`C:\Users\USERNAME\AppData\Roaming\csl` in Windows 7).  If this is left
off, scholdoc-citeproc will look for `$HOME/.csl/chicago-author-date.csl`,
and if this is not present, it will use its own version of
`chicago-author-date.csl`.

`citation-abbreviations`:  Path to a CSL abbreviations JSON file. The format
is described [here](http://citationstylist.org/2011/10/19/abbreviations-for-zotero-test-release).  Here is a short example:

    { "default": {
        "container-title": {
                "Lloyd's Law Reports": "Lloyd's Rep",
                "Estates Gazette": "EG",
                "Scots Law Times": "SLT"
        }
      }
    }

The metadata must contain either `references` or `bibliography` or
both as a source of references.  `csl` and `citation-abbreviations`
are optional.  If `csl` is not provided, `chicago-author-date.csl` will be
used by default.

## Convert mode

If the option `--bib2yaml` or `--bib2json` is used, `scholdoc-citeproc` will
not process citations; instead, it will treat its input (from stdin or
files) as a bibliography and convert it either to a scholdoc YAML metadata
section, suitable for inclusion in a scholdoc document (`--bib2yaml`), or
as a CSL JSON bibliography, suitable for import to zotero (`--bib2json`).

The `--format` option can be used to specify the bibliography format,
though when files are used, `scholdoc-citeproc` can generally guess this
from the extension.

This mode supersedes the old `biblio2yaml` program.

# OPTIONS

`-h, --help`
:   Print usage information.

`-V, --version`
:   Print version.

`-y, --bib2yaml`
:   Convert bibliography to YAML suitable for inclusion in pandoc metadata.

`-j, --bib2json`
:   Convert bibliography to CSL JSON suitable for import into Zotero.

`-f` *FORMAT*, `--format=`*FORMAT*
:   Specify format of bibliography to be converted.  Legal values are
    `biblatex`, `bibtex`, `ris`, `endnote`, `endnotexml`, `isi`,
    `medline`, `copac`, `mods`, and `json`.

# NOTES

## General

If you use a biblatex database, closely follow the specifications in the
"Database Guide" section of the biblatex manual (currently 2.8a).

If you use a CSL-YAML or CSL-JSON database, or a CSL-YAML metadata section in
your markdown document, follow the "Citation Style Language 1.0.1 Language
Specification" (<http://citationstyles.org/downloads/specification.html>).
Particularly relevant are
<http://citationstyles.org/downloads/specification.html#appendix-iii-types>
(which neither comments on usage nor specifies required and optional fields) and
<http://citationstyles.org/downloads/specification.html#appendix-iv-variables>
(which does contain comments).

## Titles: Title vs. Sentence Case

If you are using a bibtex or biblatex bibliography, then observe
the following rules:

  - English titles should be in title case.  Non-English titles should
    be in sentence case, and the `langid` field in biblatex should be
    set to the relevant language.  (The following values are treated
    as English:  `american`, `british`, `canadian`, `english`,
    `australian`, `newzealand`, `USenglish`, or `UKenglish`.)

  - As is standard with bibtex/biblatex, proper names should be
    protected with curly braces so that they won't be lowercased
    in styles that call for sentence case.  For example:

        title = {My Dinner with {Andre}}

  - In addition, words that should remain lowercase (or camelCase)
    should be protected:

        title = {Spin Wave Dispersion on the {nm} Scale}

    Though this is not necessary in bibtex/biblatex, it is necessary
    with citeproc, which stores titles internally in sentence case,
    and converts to title case in styles that require it.  Here we
    protect "nm" so that it doesn't get converted to "Nm" at this stage.

If you are using a CSL bibliography (either JSON or YAML), then observe
the following rules:

  - All titles should be in sentence case.

  - Use the `language` field for non-English titles to prevent their
    conversion to title case in styles that call for this. (Conversion
    happens only if `language` begins with `en` or is left empty.)

  - Protect words that should not be converted to title case using
    this syntax:

        Spin wave dispersion on the <span class="nocase">nm</span> scale

## Conference Papers, Published vs. Unpublished

For a formally published conference paper, use the biblatex entry type
`inproceedings` (which will be mapped to CSL `paper-conference`).

For an unpublished manuscript, use the biblatex entry type `unpublished`
without an `eventtitle` field (this entry type will be mapped to CSL
`manuscript`).

For a talk, an unpublished conference paper, or a poster presentation, use the
biblatex entry type `unpublished` with an `eventtitle` field (this entry type
will be mapped to CSL `speech`). Use the biblatex `type` field to indicate the
type, e.g. "Paper", or "Poster". `venue` and `eventdate` may be useful too,
though `eventdate` will not be rendered by most CSL styles. Note that `venue`
is for the event's venue, unlike `location` which describes the publisher's
location; do not use the latter for an unpublished conference paper.

# AUTHORS

Andrea Rossato, John MacFarlane, and Tim T.Y. Lin.

# SEE ALSO

`scholdoc` (1), `pandoc_markdown` (5).

The scholdoc-citeproc source code and all documentation may be downloaded
from <http://github.com/timtylin/scholdoc-citeproc/>.


scholdoc-citeproc.hs
module Main where
import Text.CSL.Input.Bibutils (readBiblioString, BibFormat(..))
import Text.CSL.Style (Formatted(..), Agent(..))
import Text.CSL.Reference (Reference(refId), Literal(..))
import Data.Generics ( everywhere, mkT )
import Data.List (group, sort)
import Data.Char (chr, toLower)
import Data.Monoid
import Data.Yaml
import Control.Applicative
import qualified Data.ByteString as B
import qualified Data.ByteString.Lazy.Char8 as BL8
import qualified Data.ByteString.Char8 as B8
import Data.Attoparsec.ByteString.Char8 as Attoparsec
import qualified Data.Text as T
import Data.Text.Encoding (encodeUtf8)
import Data.Aeson.Encode.Pretty (encodePretty', Config(..))
import System.Console.GetOpt
import Control.Monad
import System.IO
import System.FilePath (takeExtension)
import System.Environment (getArgs)
import System.Exit
import Data.Version (showVersion)
import Paths_scholdoc_citeproc (version)
import Text.CSL.Pandoc (processCites')
import Text.Pandoc.JSON hiding (Format)
import Text.Pandoc.Walk

main :: IO ()
main = do
  argv <- getArgs
  let (flags, args, errs) = getOpt Permute options argv
  let header = "Usage: scholdoc-citeproc [options] [file..]"
  unless (null errs) $ do
    hPutStrLn stderr $ usageInfo (unlines $ errs ++ [header]) options
    exitWith $ ExitFailure 1
  when (Version `elem` flags) $ do
    putStrLn $ "scholdoc-citeproc " ++ showVersion version
    exitWith ExitSuccess
  when (Help `elem` flags) $ do
    putStrLn $ usageInfo header options
    exitWith ExitSuccess
  if Bib2YAML `elem` flags || Bib2JSON `elem` flags
     then do
       let mbformat = case [f | Format f <- flags] of
                           [x] -> readFormat x
                           _   -> Nothing
       bibformat <- case mbformat <|>
                         msum (map formatFromExtension args) of
                         Just f   -> return f
                         Nothing  -> do
                            hPutStrLn stderr $ usageInfo
                              ("Unknown format\n" ++ header) options
                            exitWith $ ExitFailure 3
       bibstring <- case args of
                         []    -> getContents
                         xs    -> mconcat <$> mapM readFile xs
       readBiblioString bibformat bibstring >>=
         warnDuplicateKeys >>=
         if Bib2YAML `elem` flags
            then outputYamlBlock . unescapeTags . encode
            else BL8.putStrLn .
              encodePretty' Config{ confIndent = 2
                                  , confCompare = compare } .
              everywhere (mkT compressName)
     else toJSONFilter doCites

formatFromExtension :: FilePath -> Maybe BibFormat
formatFromExtension = readFormat . dropWhile (=='.') . takeExtension

readFormat :: String -> Maybe BibFormat
readFormat = go . map toLower
  where go "biblatex" = Just BibLatex
        go "bib"      = Just BibLatex
        go "bibtex"   = Just Bibtex
        go "ris"      = Just Ris
        go "endnote"  = Just Endnote
        go "enl"      = Just Endnote
        go "endnotexml" = Just EndnotXml
        go "xml"      = Just EndnotXml
        go "wos"      = Just Isi
        go "isi"      = Just Isi
        go "medline"  = Just Medline
        go "copac"    = Just Copac
        go "json"     = Just Json
        go "mods"     = Just Mods
        go _          = Nothing


doCites :: Pandoc -> IO Pandoc
doCites doc = do
  doc' <- processCites' doc
  let warnings = query findWarnings doc'
  mapM_ (hPutStrLn stderr) warnings
  return doc'

findWarnings :: Inline -> [String]
findWarnings (Span (_,["citeproc-not-found"],[("data-reference-id",ref)]) _) =
  ["scholdoc-citeproc: reference " ++ ref ++ " not found" | ref /= "*"]
findWarnings (Span (_,["citeproc-no-output"],_) _) =
  ["scholdoc-citeproc: reference with no printed form"]
findWarnings _ = []

data Option =
    Help | Version | Convert | Format String | Bib2YAML | Bib2JSON
  deriving (Ord, Eq, Show)

options :: [OptDescr Option]
options =
  [ Option ['h'] ["help"] (NoArg Help) "show usage information"
  , Option ['V'] ["version"] (NoArg Version) "show program version"
  , Option ['y'] ["bib2yaml"] (NoArg Bib2YAML) "convert bibliography to YAML"
  , Option ['j'] ["bib2json"] (NoArg Bib2JSON) "convert bibliography to JSON"
  , Option ['f'] ["format"] (ReqArg Format "FORMAT") "bibliography format"
  ]

warnDuplicateKeys :: [Reference] -> IO [Reference]
warnDuplicateKeys refs = mapM_ warnDup dupKeys >> return refs
  where warnDup k = hPutStrLn stderr $ "biblio2yaml: duplicate key " ++ k
        allKeys   = map (unLiteral . refId) refs
        dupKeys   = [x | (x:_:_) <- group (sort allKeys)]

outputYamlBlock :: B.ByteString -> IO ()
outputYamlBlock contents = do
  putStrLn "---\nreferences:"
  B.putStr contents
  putStrLn "..."

-- Compress particles and suffixes into given and last name,
-- as zotero JSON expects.  (We might also want to set parse-names = true.)
compressName :: Agent -> Agent
compressName ag = Agent{
    givenName       = gn
  , familyName      = fn
  , droppingPart    = mempty
  , nonDroppingPart = mempty
  , literal         = literal ag
  , nameSuffix      = mempty
  , commaSuffix     = False
  , parseNames      = True
  }
  where
  spcat (Formatted []) y = y
  spcat y (Formatted []) = y
  spcat x y = x <> Formatted [Space] <> y
  gnbase = givenName ag ++ [droppingPart ag | droppingPart ag /= mempty]
  gn = case (gnbase, nameSuffix ag) of
             ([], _)            -> []
             (xs, Formatted []) -> xs
             (xs, ns)           -> init xs ++
                [last xs <> Formatted [if commaSuffix ag
                                          then Str ",!"
                                          else Str ",", Space], ns]
  fn = spcat (nonDroppingPart ag) (familyName ag)

-- turn
-- id: ! "\u043F\u0443\u043D\u043A\u04423"
-- into
-- id: пункт3
unescapeTags :: B.ByteString -> B.ByteString
unescapeTags bs = case parseOnly (many $ tag <|> other) bs of
                       Left e  -> error e
                       Right r -> B.concat r

tag :: Attoparsec.Parser B.ByteString
tag = do
  _ <- string $ B8.pack ": ! "
  c <- char '\'' <|> char '"'
  cs <- manyTill (escaped c <|> other) (char c)
  return $ B8.pack ": " <> B8.singleton c <> B.concat cs <> B8.singleton c

escaped :: Char -> Attoparsec.Parser B.ByteString
escaped c = string $ B8.pack ['\\',c]

other :: Attoparsec.Parser B.ByteString
other = uchar <|> Attoparsec.takeWhile1 notspecial <|> regchar
  where notspecial = not . inClass ":!\\\"'"

uchar :: Attoparsec.Parser B.ByteString
uchar = do
  _ <- char '\\'
  num <- (2 <$ char 'x') <|> (4 <$ char 'u') <|> (8 <$ char 'U')
  cs <- count num $ satisfy $ inClass "0-9a-fA-F"
  let n = read ('0':'x':cs)
  return $ encodeUtf8 $ T.pack [chr n]

regchar :: Attoparsec.Parser B.ByteString
regchar = B8.singleton <$> anyChar
